Make RetDec compilable via MSYS2 on Windows

twitzelbos opened the pull request:
I made some changes to the cmake files that allow smooth compilation on msys2. This fixes the issue mentioned in issue 558 and another one with linking libdwarf further down the compile.

twitzelbos commented:
I'm very new to this, what is the next step to get this merged?

s3rvac commented:
Hi. Thank you for the contribution. I will review the changes and will let you know.

twitzelbos commented:
Thank you for reviewing it,

I can, by the way, reproduce the test failure on Ubuntu 18 with the regular main branch. Will look into it separately. That test failure is caused by missing linker files to pthread and dl.

s3rvac commented:
That `pthread` failure on Linux is already fixed in `master`, so no need looking into it. If you want, you can rebase your branch based on the current `master` to get all the builds to suceed.

s3rvac commented on CMakeLists.txt:
```diff
@@ -93,6 +93,11 @@ elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR
+
+	# Add some libraries needed for the MSYS build.
+	if(MSYS)
+		set(CMAKE_CXX_STANDARD_LIBRARIES "-lshlwapi -lz ${CMAKE_CXX_STANDARD_LIBRARIES}")
```
From CMake's documentation on [`CMAKE_CXX_STANDARD_LIBRARIES`](https://cmake.org/cmake/help/v3.0/variable/CMAKE_STANDARD_LIBRARIES.html):
> Libraries linked into every executable and shared library.

Are you sure that these two libraries need to be linked into **every** executable and shared library? For example, when looking which parts of RetDec use `shlwapi.h`, I have found only `retdec-utils`:
```
$ git grep shlwapi\.h
src/utils/filesystem_path.cpp:  #include <shlwapi.h>
```
The recommended way would be to add this library via `target_link_libraries()` in `src/utils/CMakeLists.txt`. Since it is already there for MSVC, I suggest simply changing
```
if(MSVC)
    target_link_libraries(retdec-utils shlwapi) # shlwapi.dll for PathRemoveFileSpec()
endif()
```
to
```
if(MSVC OR MSYS)
    target_link_libraries(retdec-utils shlwapi) # shlwapi.dll for PathRemoveFileSpec()
endif()
```

The second library is `libz` (`-lz`). AFAIK the only part that uses `libz` is LLVM. Thus, I would suggest trying adding that dependency into `deps/llvm/CMakeLists.txt` after line 226 (you can use `elseif` if you want).
```
if(MSYS)
    target_link_libraries(llvm INTERFACE debug     ${ZLIB_LIBRARIES})
    target_link_libraries(llvm INTERFACE optimized ${ZLIB_LIBRARIES})
endif()
```

Alexpux replied:
I'm primary developer of MSYS2 and I can say that condition in patch is wrong.
MSYS under CMAKE mean that you use MSYS gcc and not MINGW GCC.

Patch must be modified to:
````
if(WIN32)
    target_link_libraries(retdec-utils shlwapi) # shlwapi.dll for PathRemoveFileSpec()
endif()
````
and  replace ````if(MSYS)```` to ````if(MINGW)````

s3rvac replied:
Thank you for the comment @Alexpux.

twitzelbos replied:
Thank you all for the feedback. Will make the changes as suggested and test.

s3rvac commented:
@twitzelbos Can you please check the commits on your branch? GitHub shows 28 commits, where your commits are there twice, and there are also commits from other people. Could you please fix your `issue558` branch by taking just the most recent version of your commits and applying them to the current `master`?

twitzelbos commented:
Sorry, this must be due to by git ineptness. I have reverted one of my 4 commits and replaced with a new commit. The other 20ish commits I guess are due to the rebase with the upstream master? Is there a clean way to fix this? The diff to the upstream master should just be my patch.

Thanks, Thomas

s3rvac commented:
There are multiple ways of fixing this. One of them is as follows:
```
# 1. Ensure that you are in the root of the retdec repository:
cd retdec

# 2. Go to the master branch:
git checkout master

# 3. Generate a patch with your changes (the diff between your branch and master):
git diff master..issue558 > patch.txt

# 4. Delete the original local branch.
git branch -d issue558

# 5. Create a new branch off the current master and switch to it.
git checkout -b issue558

# 6. Apply changes from the saved patch:
git apply patch.txt

# 7. Stage the changed files:
git add deps/libdwarf/CMakeLists.txt deps/llvm/CMakeLists.txt deps/yaracpp/CMakeLists.txt src/utils/CMakeLists.txt

# 8. Commit them in a single commit (an editor will show up):
git commit

# 9. Remove the no-longer-needed patch:
rm patch.txt

# 10. Push the changes to the original branch on GitHub (--force is required as we are overwriting history):
git push --force origin issue558
```
After this, there should be only a single commit in this PR.

twitzelbos commented:
I can do this, but if I may use some more of your time, I like to learn why this has to be done. My branch, as it is right now, has this diff:
`git diff upstream/master..issue558
diff --git a/deps/libdwarf/CMakeLists.txt b/deps/libdwarf/CMakeLists.txt
index 3839a4fe..a3a7b068 100644
--- a/deps/libdwarf/CMakeLists.txt
+++ b/deps/libdwarf/CMakeLists.txt
@@ -71,6 +71,9 @@ target_include_directories(libdwarf SYSTEM INTERFACE ${source_dir}/libdwarf)
 if(MSVC)
        target_link_libraries(libdwarf INTERFACE ${binary_dir}/libdwarf/libdwarf/$<CONFIGURATION>/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libdwarf.lib)
        target_link_libraries(libdwarf INTERFACE ${binary_dir}/libelf/lib/$<CONFIGURATION>/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libelf.lib)
+elseif(MSYS)
+       target_link_libraries(libdwarf INTERFACE ${LIBDWARF_INSTALL_DIR}/bin/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libdwarf${CMAKE_SHARED_LIBRARY_SUFFIX})
+       target_link_libraries(libdwarf INTERFACE ${LIBDWARF_INSTALL_DIR}/bin/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libelf${CMAKE_SHARED_LIBRARY_SUFFIX})
 else() # Linux
        target_link_libraries(libdwarf INTERFACE ${LIBDWARF_INSTALL_DIR}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libdwarf${CMAKE_SHARED_LIBRARY_SUFFIX})
        target_link_libraries(libdwarf INTERFACE ${LIBDWARF_INSTALL_DIR}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}retdec-libelf${CMAKE_SHARED_LIBRARY_SUFFIX})
@@ -78,6 +81,8 @@ endif()

 if(MSVC)
        install(DIRECTORY "${LIBDWARF_INSTALL_DIR}/bin" DESTINATION "${CMAKE_INSTALL_PREFIX}")
+elseif(MSYS)
+       install(DIRECTORY "${LIBDWARF_INSTALL_DIR}/bin" DESTINATION "${CMAKE_INSTALL_PREFIX}")
 else() # Linux
        install(DIRECTORY "${LIBDWARF_INSTALL_DIR}/lib" DESTINATION "${CMAKE_INSTALL_PREFIX}")
 endif()
diff --git a/deps/llvm/CMakeLists.txt b/deps/llvm/CMakeLists.txt
index 2794fed6..9869cd8d 100644
--- a/deps/llvm/CMakeLists.txt
+++ b/deps/llvm/CMakeLists.txt
@@ -1,5 +1,5 @@
 find_package(Threads REQUIRED)
-if(UNIX)
+if(UNIX OR MINGW)
        find_package(ZLIB REQUIRED)
 endif()

@@ -223,6 +223,9 @@ if(UNIX)
        endif()
        target_link_libraries(llvm INTERFACE debug     ${ZLIB_LIBRARIES} ${CMAKE_DL_LIBS} ${EXECINFO})
        target_link_libraries(llvm INTERFACE optimized ${ZLIB_LIBRARIES} ${CMAKE_DL_LIBS} ${EXECINFO})
+elseif(MINGW)
+       target_link_libraries(llvm INTERFACE debug     ${ZLIB_LIBRARIES})
+       target_link_libraries(llvm INTERFACE optimized ${ZLIB_LIBRARIES})
 endif()

 # Set include directories.
diff --git a/deps/yaracpp/CMakeLists.txt b/deps/yaracpp/CMakeLists.txt
index e086b20a..4486088b 100644
--- a/deps/yaracpp/CMakeLists.txt
+++ b/deps/yaracpp/CMakeLists.txt
@@ -78,6 +78,9 @@ if(MSVC)
        set(YARA_WIN_DIR       ${YARA_DIR}/windows/vs2015)
        set(YARA_LIBRARY_DIR   ${YARA_WIN_DIR}/libyara/$<CONFIGURATION>)
        set(YARAC_PATH         ${YARA_WIN_DIR}/\${CMAKE_INSTALL_CONFIG_NAME}/${YARAC_NAME})
+elseif(MSYS)
+       set(YARA_LIBRARY_DIR   ${YARA_DIR}/libyara/.libs)
+       set(YARAC_PATH         ${YARA_DIR}/yarac.exe)
 else()
        set(YARA_LIBRARY_DIR   ${YARA_DIR}/libyara/.libs)
        set(YARAC_PATH         ${YARA_DIR}/yarac)
diff --git a/src/utils/CMakeLists.txt b/src/utils/CMakeLists.txt
index 642c9187..d879f306 100644
--- a/src/utils/CMakeLists.txt
+++ b/src/utils/CMakeLists.txt
@@ -16,7 +16,7 @@ set(RETDEC_UTILS_SOURCES

 add_library(retdec-utils STATIC ${RETDEC_UTILS_SOURCES})
 target_link_libraries(retdec-utils whereami)
-if(MSVC)
+if(WIN32)
        target_link_libraries(retdec-utils shlwapi) # shlwapi.dll for PathRemoveFileSpec()
 endif()
 target_link_libraries(retdec-utils mpark_variant)`

Are you worried about the clutter in the history?

s3rvac commented:
> I can do this, but if I may use some more of your time, I like to learn why this has to be done.

Sure. When you click on [Files changed](https://github.com/avast/retdec/pull/606/files), there are changes in 15 files, and most of them are not related to the present PR in any way. This causes unnecessary difficulties to both me (as a reviewer) and future readers of the PR. The reason is that GitHub shows the changes of all new commits in the branch rather than changes when compared to `master`.

> Are you worried about the clutter in the history?

No. I would squash all the commits from the branch into a single one when merging anyway, so there won't be any clutter.

twitzelbos commented:
Okay, will do as you suggested. It just doesn't seem like that's the intended way with git.

s3rvac commented:
Thank you, the changes are now shown correctly :+1: I will review the code, and if everything works as expected, I will merge the PR.
